fortios/docker/launch.py

```diff
@@ -0,0 +1,199 @@
+#!/usr/bin/env python3
+import datetime
+import logging
+import os
+import re
+import signal
+import sys
+import uuid
+
+import vrnetlab
+
+
+def handle_SIGCHLD(_unused_signal, _unused_frame):
+    os.waitpid(-1, os.WNOHANG)
+
+
+def handle_SIGTERM(_unused_signal, _unused_frame):
+    sys.exit(0)
+
+
+signal.signal(signal.SIGINT, handle_SIGTERM)
+signal.signal(signal.SIGTERM, handle_SIGTERM)
+signal.signal(signal.SIGCHLD, handle_SIGCHLD)
+
+TRACE_LEVEL_NUM = 9
+logging.addLevelName(TRACE_LEVEL_NUM, "TRACE")
+
+
+def trace(self, message, *args, **kws):
+    # Yes, logger takes its '*args' as 'args'.
+    if self.isEnabledFor(TRACE_LEVEL_NUM):
+        self.log(TRACE_LEVEL_NUM, message, *args, **kws)
+
+
+logging.Logger.trace = trace
+
+
+class FortiOS_vm(vrnetlab.VM):
+    def __init__(self, hostname, username, password, conn_mode):
+        for e in os.listdir("."):
+            if re.search(".qcow2$", e):
+                disk_image = "./" + e
+        # call parents __init__ function here
+        super(FortiOS_vm, self).__init__(
+            username,
+            password,
+            disk_image=disk_image,
+            ram=2048,
+            driveif="virtio",
+            # fortios fails to respond to network requests if the pci bus is setup :D
+            provision_pci_bus=False,
+        )
+        self.conn_mode = conn_mode
+        self.hostname = hostname
+        self.num_nics = 12
+        self.nic_type = "virtio-net-pci"
+        self.highest_port = 0
+        self.qemu_args.extend(["-uuid", str(uuid.uuid4())])
+        self.qemu_args.extend(["-smp", "1", "-cpu", "host"])
+        self.spins = 0
+        self.running = None
+
+        # set up the extra empty disk image
+        # for fortigate logs
+        vrnetlab.run_command(
+            ["qemu-img", "create", "-f", "qcow2", "empty.qcow2", "30G"]
+        )
+
+        self.qemu_args.extend(
+            [
+                "-drive",
+                "if=virtio,format=qcow2,file=empty.qcow2,index=1",
+            ]
+        )
+
+    def gen_nics(self):
+        """Generate qemu args for the normal traffic carrying interface(s)"""
+        res = super(FortiOS_vm, self).gen_nics()
+
+        if res:
+            netdev_list = list(filter(lambda ports: "netdev=" in ports, res))
+
+            port_list = []
+
+            for netdev in netdev_list:
+                port_list.append(int(netdev.split(",")[1][-2:]))
+
+            self.highest_port = max(port_list)
+
+            return res
+
+        else:
+            return res
```

Comment:
@robotwalk what exactly happens here? It seems this is smth you added specifically for fortigate

Reply:
Nevermind about this. It was just added to get the next free port number available and configure something on the device itself but on another interface


By default just port1 is available
```
vr-fortinet # show system interface
config system interface
    edit "port1"
        set vdom "root"
        set mode dhcp
        set allowaccess ping https ssh fgfm
        set type physical
        set snmp-index 1
    next
 ``` 

if port2 is edited


e.g.
`self.wait_write("edit port%s" % (self.highest_port + 1)) `
`self.wait_write("set vdom root"`
etc.

would result in 
```
    edit "port2"
        set vdom "root"
        set allowaccess ping https ssh
        set type loopback
        set snmp-index 4
    next
```
included in the `show system interface`

I guess we can just remove this since it was just a special case at work to set some ip addresses on specific interfaces.